airbyte-workers/src/main/java/io/airbyte/workers/protocols/airbyte/AirbyteMessageTracker.java

```diff
@@ -4,51 +4,234 @@
+    this.totalRecordsEmittedByStream = new HashMap<>();
+    this.stateDeltaTracker = stateDeltaTracker;
+    this.nextStreamIndex = 0;
+    this.unreliableCommittedCounts = false;
   }
 
   @Override
-  public void accept(final AirbyteMessage message) {
-    if (message.getType() == AirbyteMessage.Type.RECORD) {
-      recordCount.incrementAndGet();
-      // todo (cgardens) - pretty wasteful to do an extra serialization just to get size.
-      numBytes.addAndGet(Jsons.serialize(message.getRecord().getData()).getBytes(Charsets.UTF_8).length);
+  public void acceptFromSource(final AirbyteMessage message) {
+    switch (message.getType()) {
+      case RECORD -> handleSourceEmittedRecord(message.getRecord());
+      case STATE -> handleSourceEmittedState(message.getState());
+      default -> log.warn("Invalid message type for message: {}", message);
+    }
+  }
+
+  @Override
+  public void acceptFromDestination(final AirbyteMessage message) {
+    switch (message.getType()) {
+      case STATE -> handleDestinationEmittedState(message.getState());
+      default -> log.warn("Invalid message type for message: {}", message);
+    }
+  }
+
+  /**
+   * When a source emits a record, increment the running record count, the total record count, and the
+   * total byte count for the record's stream.
+   */
+  private void handleSourceEmittedRecord(final AirbyteRecordMessage recordMessage) {
+    final short streamIndex = getStreamIndex(recordMessage.getStream());
+
+    final long currentRunningCount = runningCountByStream.getOrDefault(streamIndex, 0L);
+    this.runningCountByStream.put(streamIndex, currentRunningCount + 1);
+
+    final long currentTotalCount = totalRecordsEmittedByStream.getOrDefault(streamIndex, 0L);
+    this.totalRecordsEmittedByStream.put(streamIndex, currentTotalCount + 1);
+
+    // todo (cgardens) - pretty wasteful to do an extra serialization just to get size.
+    final int numBytes = Jsons.serialize(recordMessage.getData()).getBytes(Charsets.UTF_8).length;
+    final long currentTotalStreamBytes = totalBytesEmittedByStream.getOrDefault(streamIndex, 0L);
+    this.totalBytesEmittedByStream.put(streamIndex, currentTotalStreamBytes + numBytes);
+  }
+
+  /**
+   * When a source emits a state, persist the current running count per stream to the
+   * {@link StateDeltaTracker}. Then, reset the running count per stream so that new counts can start
+   * recording for the next state. Also add the state to list so that state order is tracked
+   * correctly.
+   */
+  private void handleSourceEmittedState(final AirbyteStateMessage stateMessage) {
+    sourceOutputState.set(new State().withState(stateMessage.getData()));
+    totalEmittedStateMessages.incrementAndGet();
+    final int stateHash = getStateHashCode(stateMessage);
+    try {
+      if (!unreliableCommittedCounts) {
+        stateDeltaTracker.addState(stateHash, runningCountByStream);
+      }
+    } catch (final CapacityExceededException e) {
```

Comment:
I considered doing that and was torn between the two. I went for the checked exception route because it does feel like an 'exceptional' case that callers should be forced to handle explicitly. 

If I went for the boolean instead of the checked exception, I think I'd need `commitStateHash` to return a boolean as well. Right now it throws the same exception if capacity was exceeded earlier.

Not super opinionated either way, and my last team went kind of crazy with checked exceptions everywhere so I'm happy to go with the boolean if that feels more consistent with our team conventions!

Reply:
legit. i'm fine with it is as it is